src/cholesky.jl

```diff
@@ -1,54 +1,118 @@
-# Generic Cholesky decomposition for fixed-size matrices, mostly unrolled
-@inline function Base.chol(A::StaticMatrix)
-    ishermitian(A) || Base.LinAlg.non_hermitian_error("chol")
-    _chol(Size(A), A)
-end
+import Base: chol, cholfact, \
 
-@inline function Base.chol(A::Base.LinAlg.RealHermSymComplexHerm{<:Real, <:StaticMatrix})
-    _chol(Size(A), A.data)
-end
+@inline _chol(A::StaticMatrix{<:Any,<:Any,T}, ::Type{UpperTriangular}) where {T} = _chol(Size(A), A, UpperTriangular)
+@inline _chol(A::StaticMatrix{<:Any,<:Any,T}, ::Type{LowerTriangular}) where {T} = _chol(Size(A), A, LowerTriangular)
 
-@generated function _chol(::Size{(1,1)}, A::StaticMatrix)
-    @assert size(A) == (1,1)
-    T = promote_type(typeof(sqrt(one(eltype(A)))), Float32)
-    newtype = similar_type(A,T)
+@generated function _chol(::Size{s}, A::StaticMatrix{<:Any,<:Any,T}, ::Type{UpperTriangular}) where {s, T}
+    if s[1] != s[2]
+        error("matrix must be square")
```

Comment:
This check is not necessary; the `ishermitian` check does this before we enter this function.